### ISSUE TITLE
Unable to publish package (500 Internal Server Error)

### ISSUE THREAD

**Issue by mreid-tt** [SynoCommunity/spkrepo#95]
When attempting to publish a package the following is reported:

```
HTTP/1.1 500 Internal Server Error
Connection: keep-alive
Content-Length: 141
Content-Type: text/html
Date: Sun, 23 Apr 2023 12:02:35 GMT
Server: nginx/1.23.2

<html>
  <head>
    <title>Internal Server Error</title>
  </head>
  <body>
    <h1><p>Internal Server Error</p></h1>
    
  </body>
</html>
```

This appears to be happening consistently for moderately sized SPK files (50-75 MB) and does not appear to be based on time of day (server load).

**Comment by mreid-tt** [comment]
@Diaoul, @ymartin59, any thoughts on this?

**Comment by mreid-tt** [comment]
This leaves the question of the original issue of the uploaded file from the stream not being present at the expected path during publishing. From examining the server code I see that publishing involves:

1. When a new package is posted, the server creates an SPK object. This object is based on the SPK class and is passed the request data as its argument
2. The server creates a Package object if one doesn't already exist, based on the package name in the request
3. The server parses the SPK content to extract the necessary information such as firmware, package name, version, etc.
4. The server then creates a Version object for the package, based on the package name and version number in the request. If a Version object with the same version number already exists, the new version is not created
5. The server also creates a Build object, which is not active by default. If a Build object with the same architectures exists, a conflict error is returned
6. The server saves the files by (a) creating a new directory for the package if there is a new Package object, (b) creating a new version sub-directory for the package if there is a new Version object and (c) saves the named spk file (using the SPK object) to this version sub-directory

Now the error seen: `FileNotFoundError: [Errno 2] No such file or directory: '/data/lidarr/11/` suggests that in attempting to save the named spk file, it is unable to do so. It is possible that while the path '/data/lidarr/' does exist (since there are other versions present), the version sub-directory is not. A missing sub-directory may cause the save to fail.

**Comment by mreid-tt** [comment]
I suspect that there may have been some sort of corruption when the first package from lidarr v11 was uploaded. The admin interface reports the following:

<img width="1186" alt="Screenshot 2023-04-25 at 1 03 26 PM" src="https://user-images.githubusercontent.com/943378/234358500-b4ef5a21-1627-4b05-bed5-7b908d869253.png">

However, when we look at the webpage it shows the following:

<img width="365" alt="Screenshot 2023-04-25 at 1 42 32 PM" src="https://user-images.githubusercontent.com/943378/234358716-a4504a21-4451-4352-ac72-a4599f93a567.png">

This version should not be visible since it is disabled in the admin interface. The application icon in the webpage was also not visible (before v12 was uploaded), which suggests that the code which runs when the new Version object is present was not initially run.

**Comment by mreid-tt** [comment]
I was publishing some other packages and I see the same corruption happening with prowlarr v6. The admin interface reports the following:

<img width="1179" alt="Screenshot 2023-04-26 at 7 53 47 AM" src="https://user-images.githubusercontent.com/943378/234567228-2f904954-7ad7-4304-8e40-dfe7ac9ad90b.png">

However, the webpage at `https://synocommunity.com/package/prowlarr` shows the following:

<img width="1172" alt="Screenshot 2023-04-26 at 7 54 02 AM" src="https://user-images.githubusercontent.com/943378/234567343-9557817f-fec5-41cf-bf11-89729317e015.png">

Note the missing icon as I indicated above with lidarr v11 when no newer version is present. I believe this happens because this code is not being executed:

https://github.com/SynoCommunity/spkrepo/blob/76951e0e5850504b16c9164b97a23883488269d6/spkrepo/views/api.py#L260-L263

**Comment by Diaoul** [comment]
Yes there must've been a failure at a time that did create the version but not the appropriate directories, causing subsequent ones to fail.
Maybe the best move is to delete the version and re-upload everything?

**Comment by mreid-tt** [comment]
> Yes there must've been a failure at a time that did create the version but not the appropriate directories, causing subsequent ones to fail.
> Maybe the best move is to delete the version and re-upload everything?

Thanks for your feedback. Before deleting the version can you check:

1. The application log to see if it is reporting the same `FileNotFoundError: [Errno 2] No such file or directory`  errors
2. The file system to determine if the path `/data/prowlarr/6/` exists. If so, what’s in that directory?

if this is the case then I’d request that the prowlarr v6 be deleted. I don’t think I have permissions to do that.

**Comment by mreid-tt** [comment]
hey @Diaoul, any feedback on the above?

In the meantime, would you be so kind as to delete this version:

<img width="1178" alt="Screenshot 2023-04-27 at 7 09 56 AM" src="https://user-images.githubusercontent.com/943378/234845224-4df8fa32-ed17-40fb-8c25-f3c4d71c44d6.png">

**Comment by publicarray** [comment]
@mreid-tt I have deleted lidarr v11 for you

**Comment by mreid-tt** [comment]
thanks @publicarray, I was hoping we could check the application logs and the file system for prowlarr to confirm the theory above. But if this is a hassle we can just delete the problematic version:

<img width="1177" alt="Screenshot 2023-04-27 at 7 52 36 AM" src="https://user-images.githubusercontent.com/943378/234854176-c1304749-c82a-4348-bfa2-57607dfa8056.png">

**Comment by publicarray** [comment]
@mreid-tt the folder was never created...

![image](https://user-images.githubusercontent.com/5497998/234856045-d5499086-63d0-4338-8c0c-2181ea3094e6.png)

```
publicarray@vps-b34822e4 /e/synocommunity> ls data/prowlarr
4/  5/

```

**Comment by mreid-tt** [comment]
> @mreid-tt the folder was never created...
> 
> ```
> publicarray@vps-b34822e4 /e/synocommunity> ls data/prowlarr
> 4/  5/
> ```

Thanks for the confirmation.

> I have deleted the broken version

Great! I'll proceed with a fresh build and upload.

**Comment by mreid-tt** [comment]
> We had problems uploading before (due to load) but never like this (missing folders/files)...

Hmm, okay. Perhaps a possible fix would be to adjust the code to verify that a folder is created (for both the package and version functions) before committing a new entry to the database.

**Comment by publicarray** [comment]
my2 cents but maybe only upload one file at a time. IIRC phython/postgresql are bad at working with multiple upload requests at the same time. This smells like a race condition

**Comment by mreid-tt** [comment]
> I have deleted the broken version

Can you confirm this was done? I'm still seeing this:

<img width="1174" alt="Screenshot 2023-04-27 at 8 12 29 AM" src="https://user-images.githubusercontent.com/943378/234858349-f8dc5125-fb05-4594-99a2-c2280461458e.png">

**Comment by publicarray** [comment]
hm maybe build them without publishing ongithub and download the result. Then manually create the upload request with httpie. I know it's a workaround but we can also test if you for example upload all packages at once if this courses this problem again.

**Comment by publicarray** [comment]
Maybe your page is cached.
![image](https://user-images.githubusercontent.com/5497998/234859069-e40a2a59-bf0f-4058-9d20-2c2d801db466.png)

**Comment by publicarray** [comment]
we got one :)

```
publicarray@vps-b34822e4 /e/synocommunity> ls -l data/prowlarr/6
total 55524
-rw-r--r-- 1 root root    44917 Apr 27 12:25  icon_256.png
-rw-r--r-- 1 root root     7211 Apr 27 12:25  icon_72.png
-rw-r--r-- 1 root root 56801280 Apr 27 12:25 'prowlarr.v6.f15047[apollolake-avoton-braswell-broadwell-broadwellnk-broadwellntbap-bromolow-cedarview-denverton-dockerx64-geminilake-grantley-purley-kvmx64-v1000-x86-x86_64].spk'
```

**Comment by mreid-tt** [comment]
Hi @Diaoul, now that we have conclusively proven that the issue is rooted in the version folder not being created, this suggests that this code block is not executed:

https://github.com/SynoCommunity/spkrepo/blob/66c4ac66af3b0e8e9adc1c9ef92df3b599baab73/spkrepo/views/api.py#L260-L263

If this is the case, how could we make the function more resilient? Should the `os.mkdir()` be put into a try block to catch any exceptions? Should we have a line right afterwards to verify that the path exists? The worry I have is that it thinks the command was successful. This is further confirmed in the log which reports the following back in GitHub:

```
  Package published successfully
  {"package": "prowlarr", "version": "20230426-6", "firmware": "6.1-15047", "architectures": ["evansport"]}
```

Why might this be the case? Are the filesystem actions for directory creation and `build.save()` being executed asynchronously and never complete even though the function believes it was executed? For the sequence of events, when exactly is the new version added to the database?
If it is not asynchronous, are multiple function calls trying to create the same version when GitHub runs parallel build operations? If so, why wouldn't at least one function call be completed? That way even if the other parallel calls fail (or do nothing because the folder exists), the successful one should create the folder, shouldn't it?

Interested in your thoughts before I create a PR so that the solution is effective.

**Comment by publicarray** [comment]
My guess is that if a version exist in the db, but the transaction isn't finished yet and another one is created at the same time we get this condition. Not sure why not at least one mkdir command is executed though. Maybe both transactions exit early because a version exists already?

**Comment by publicarray** [comment]
I also noted that if there is an exception files are removed so what happens when a file is created and at the same time removed by a different request?

**Comment by mreid-tt** [comment]
> I also noted that if there is an exception, files are removed so what happens when a file is created and at the same time removed by a different request?

Ahh! That's a good point, you are referring to this part of the code:

https://github.com/SynoCommunity/spkrepo/blob/66c4ac66af3b0e8e9adc1c9ef92df3b599baab73/spkrepo/views/api.py#L265-L272

So, it's possible it was created just fine but there was some sort of exception (perhaps even in creating the icons or saving the build), then the entire version folder would be dumped. If this is the case then the version should also be removed from the database (or not added if this is done later). Need to better understand the database interactions to confirm.

**Comment by mreid-tt** [comment]
So I shared our problem with ChatGPT and it suggested something like this:

```
import threading

# create a lock to ensure that only one request can create a new version at a time
version_lock = threading.Lock()

def create_new_version(package, version_number):
    create_version = False
    with version_lock:
        version = Version.query.filter_by(package_id=package.id, version=version_number).first()
        if version is None:
            version = Version(package=package, version=version_number)
            db.session.add(version)
            db.session.commit()
            create_version = True
    if create_version:
        # create the directory for the new version
        version_path = os.path.join(current_app.config['DATA_PATH'], package.name, str(version_number))
        os.makedirs(version_path, exist_ok=True)
    return version
```

It's a different structure from the code currently there but may serve as inspiration for a possible solution.

**Comment by Diaoul** [comment]
Thinking a bit about this, I am not sure why we need a clean up the directory structure :shrug: We could very well make sure the path always exists.
However, this could hide another consistency problem (like in #100)

**Comment by publicarray** [comment]
I think I agree with @Diaoul here, using locks just overcomplicates things. I think just checking if there is an exception like "version already exists" and continuing the code anyway rather than cleaning up the folders is enough in this case. I'd like to mention that what I think is happening is a hypothesis and not proven yet.

There could have been an issue with the redis or database connection. Unfortunately, there aren't time codes in the app logs. 

```
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
 	Is the server running on host "db" (172.18.0.3) and accepting
 	TCP/IP connections on port 5432?

redis.exceptions.ConnectionError: Error while reading from redis:6379 : (104, 'Connection reset by peer')
Exception possibly due to cache backend.
```

**Comment by mreid-tt** [comment]
Thanks for the additional perspectives. Taking a step back the issue is really about the file system and not the database. Re-looking at the code we have these operations to create folders:

https://github.com/SynoCommunity/spkrepo/blob/66c4ac66af3b0e8e9adc1c9ef92df3b599baab73/spkrepo/views/api.py#L259

My understanding is that if the folder already exists, it will throw a `FileExistsError`. This is another possible reason for the try block failing and a more likely scenario for multiple workers adding a new build. I've done another PR with a far simpler fix for consideration in #101.

**Comment by mreid-tt** [comment]
> Thinking a bit about this, I am not sure why we need a clean up the directory structure 🤷 We could very well make sure the path always exists.

I actually like the elegance of cleaning up the structure on an exception so you maintain a clean filesystem. I mean we can adjust the code to something like this:

```
if create_package:
    dir_path = os.path.join(data_path, package.name)
    if not os.path.exists(dir_path):
        os.mkdir(dir_path)
```

But I really believe now that the exception on the try block is more likely caused by the folder already existing which can easily be mitigated.

**Comment by mreid-tt** [comment]
@publicarray, looking back at the application logs you kindly provided, I believe this new theory is correct. In the logs we first see this error:

```
[2023-04-16 12:32:53 +0000] [7] [ERROR] Error handling request /packages
Traceback (most recent call last):
  File "/usr/src/app/spkrepo/views/api.py", line 261, in post
    os.mkdir(os.path.join(data_path, package.name, str(version.version)))
FileExistsError: [Errno 17] File exists: '/data/lidarr/11'
```

Immediately after that we have these errors and similar:

```
[2023-04-16 12:33:40 +0000] [10] [ERROR] Error handling request /packages
Traceback (most recent call last):
  File "/usr/src/app/spkrepo/views/api.py", line 264, in post
    build.save(spk.stream)
  File "/usr/src/app/spkrepo/models.py", line 357, in save
    with io.open(
FileNotFoundError: [Errno 2] No such file or directory: '/data/lidarr/11/lidarr.v11.f42661[comcerto2k].spk'
```

I believe this supports the theory that:
1. Two workers attempt to create a directory at '/data/lidarr/11'
2. One succeeds and the other fails
3. The failing one deletes the directory path
4. Other workers saving builds fail because the path does not exist

**Comment by Diaoul** [comment]
I am not sure why we would end up publishing multiple packages at the same time, maybe we can fix that instead? The code is full of race conditions and absolutely not ready to handle parallel uploads of the same packages.

Also worth noting that it seems it's the first time it occurs in close to 8 years of this code being in production. Unless we're confident the code will not break elsewhere when introducing a fix, I would be OK with a one-time procedure to manually fix this and document it somewhere :grimacing:

**Comment by mreid-tt** [comment]
> I am not sure why we would end up publishing multiple packages at the same time, maybe we can fix that instead? The code is full of race conditions and absolutely not ready to handle parallel uploads of the same packages.

From my understanding, I believe we have been doing this for over two years based on the age of the wiki page... [Publish a package from Github](https://github.com/SynoCommunity/spksrc/wiki/Publish-a-package-from-Github). To my knowledge there isn't a way to tell the GitHub workflow to not run in parallel. As such, once we use it, there is the potential to be sending multiple builds.

> Also worth noting that it seems it's the first time it occurs in close to 8 years (I've lost count) of this code being in production. Unless we're confident the code will not break elsewhere when introducing a fix, I would be OK with a one-time procedure to manually fix this and document it somewhere 😬

I understand the hesitation in making such a fundamental change but I'd like to share another perspective:

1. I don't believe that this is the first time this has happened. In the application logs shared with me that had entries between 2023-02-19 and 2023-04-25 there was one other version directory missing '/data/owncloud/9'. More likely what has been happening  is that because the GitHub publish function did not check for errors (which it now does as part of [#5701](https://github.com/SynoCommunity/spksrc/pull/5701)) the uploads appeared to be successful when they were not (see https://github.com/SynoCommunity/spksrc/pull/5619#issuecomment-1528923235)
2. You are correct that there is the possibility of the code breaking elsewhere but from a database perspective you have referential integrity constraints built-in to the Package and Version models so should there be a duplicate addition, it would cleanly fail. For the addition of the Build, there is the possibility of the object being linked to the incorrect Version object and I can't think of a way to address this without locks
3. As far as the filesystem, if the folder already exists we should be fine with the proposed fix. If we try to save duplicate icons, based on the statement construction for `icon.save()` the existing ones should be overwritten which should also be okay
4. As far as manually fixing it, the way I've seen it done is to delete the version in the admin panel and re-upload from scratch. While this can work, not all devs have admin permissions to do this. Additionally, it does not prevent it from happening again with another GitHub workflow (unless you manually re-run each failed job one at a time)

**Comment by Diaoul** [comment]
Thanks for the additional context, it makes sense to try with this then :+1:

**Comment by publicarray** [comment]
For some history. The first Ci used travis and circleCI processed one arch at a time. Also it was expected that every dev build on their own devices before publishing so that they would test the build locally as well (that hasn't changed) but with GitHub actions it became easier and we implemented the publishing part as well. So if you build locally and test the one or two architectures you don't need to build all of them to publish. Lowering the barrier a bit.
